substrate/frame/asset-rewards/src/lib.rs

```diff
@@ -0,0 +1,718 @@
+				pool_info.reward_asset_id,
+				&Self::pool_account_id(&pool_id)?,
+				&dest,
+				amount,
+				Preservation::Preserve,
+			)?;
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Derive a pool account ID from the pallet's ID.
+		pub fn pool_account_id(id: &PoolId) -> Result<T::AccountId, DispatchError> {
+			if Pools::<T>::contains_key(id) {
+				Ok(T::PalletId::get().into_sub_account_truncating(id))
+			} else {
+				Err(Error::<T>::NonExistentPool.into())
+			}
+		}
+
+		/// Computes update pool and staker reward state.
+		///
+		/// Should be called prior to any operation involving a staker.
+		///
+		/// Returns the updated pool and staker info.
+		///
+		/// NOTE: this function has no side-effects. Side-effects such as storage modifications are
+		/// the resopnsibility of the caller.
+		pub fn update_pool_and_staker_rewards(
+			pool_info: &PoolInfoFor<T>,
+			staker_info: &PoolStakerInfo<T::Balance>,
+		) -> Result<(PoolInfoFor<T>, PoolStakerInfo<T::Balance>), DispatchError> {
+			let pool_info = Self::update_pool_rewards(pool_info)?;
+
+			let mut new_staker_info = staker_info.clone();
+			new_staker_info.rewards = Self::derive_rewards(&pool_info, &staker_info)?;
+			new_staker_info.reward_per_token_paid = pool_info.reward_per_token_stored;
+			return Ok((pool_info, new_staker_info));
+		}
+
+		/// Computes update pool reward state.
+		///
+		/// Should be called every time the pool is adjusted, and a staker is not involved.
+		///
+		/// Returns the updated pool and staker info.
+		///
+		/// NOTE: this function has no side-effects. Side-effects such as storage modifications are
+		/// the resopnsibility of the caller.
+		pub fn update_pool_rewards(
+			pool_info: &PoolInfoFor<T>,
+		) -> Result<PoolInfoFor<T>, DispatchError> {
+			let reward_per_token = Self::reward_per_token(&pool_info)?;
+
+			let mut new_pool_info = pool_info.clone();
+			new_pool_info.last_update_block = frame_system::Pallet::<T>::block_number();
+			new_pool_info.reward_per_token_stored = reward_per_token;
+
+			Ok(new_pool_info)
+		}
+
+		/// Derives the current reward per token for this pool.
+		///
+		/// This is a helper function for `update_pool_rewards` and should not be called directly.
+		fn reward_per_token(pool_info: &PoolInfoFor<T>) -> Result<T::Balance, DispatchError> {
+			if pool_info.total_tokens_staked.is_zero() {
+				return Ok(pool_info.reward_per_token_stored)
+			}
+
+			let rewardable_blocks_elapsed: u32 =
+				match Self::last_block_reward_applicable(pool_info.expiry_block)
+					.saturating_sub(pool_info.last_update_block)
+					.try_into()
+				{
+					Ok(b) => b,
+					Err(_) => return Err(Error::<T>::BlockNumberConversionError.into()),
+				};
+
+			Ok(pool_info.reward_per_token_stored.saturating_add(
+				pool_info
+					.reward_rate_per_block
+					.saturating_mul(rewardable_blocks_elapsed.into())
+					.saturating_mul(PRECISION_SCALING_FACTOR.into())
+					.ensure_div(pool_info.total_tokens_staked)?,
```

Comment:
Why would it not use our fractional types like `FixedU64`?

Reply:
Yea, the `PRECISION_SCALING_FACTOR` seems a bit odd IMHO.

Reply:
It's a common pattern on EVM, but yes I guess not necessary here. I'll update to use FixedU64.

Reply:
Hmm I'm not sure actually this is best, because all other units are just stored as a Balance. 

If we have the stored rewards as a `Fixed`, we need to introduce a whole new set of fail-able conversions and handling of decimals that's not required if we just make everything a balance.

I was able to reduce the scaling factor to just 4096, with minimal rounding errs. I feel that is fine.